// cdk-dynamo-table-viewer/lib/table-viewer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cdk = require("@aws-cdk/core");
const apigw = require("@aws-cdk/aws-apigateway");
const lambda = require("@aws-cdk/aws-lambda");
const path = require("path");
/**
 * Installs an endpoint in your stack that allows users to view the contents
 * of a DynamoDB table through their browser.
 */
class TableViewer extends cdk.Construct {
    constructor(parent, id, props) {
        super(parent, id);
        const handler = new lambda.Function(this, 'Rendered', {
            code: lambda.Code.asset(path.join(__dirname, 'lambda')),
            runtime: lambda.Runtime.NODEJS_12_X,
            handler: 'index.handler',
            environment: {
                TABLE_NAME: props.table.tableName,
                TITLE: props.title || '',
                SORT_BY: props.sortBy || ''
            }
        });
        props.table.grantReadData(handler);
        const home = new apigw.LambdaRestApi(this, 'ViewerEndpoint', { handler });
        this.endpoint = home.url;
    }
}
exports.TableViewer = TableViewer;
